### DISCUSSION TITLE
Storage: Upload to Bucket from Url

### DISCUSSION THREAD

**Discussion by psteinroe** [supabase/supabase#22590] [Feature Requests]
Hey,

we have a lot of cased in our backend where we need to transfer a file from an external api to our storage. In almost all cases, the file is provided via an url. We download it from the endpoint in memory, and upload it to supabase again. It would be nice if the storage api could expose an endpoint that allows uploading a file by passing an url:

1. User passes url
`supabase.storage.from('bucket').upload(filePath, 'https://some-url');`

2. Storage Api makes HEAD request to get metadata of file
Check file size and content type returned by HEAD request against bucket config and throw if invalid 

3. Storage Api downloads file from url directly to bucket at given `filePath` 

Thanks!

**Comment by lucasmonstrox** [comment]
+1

**Comment by alexwhitmore** [comment]
Hi there,

Not exactly a native solution, but I'm working on an npm package (soon to be @supabase-community/storage-helpers) that provides `uploadFromUrl` functionality via an Edge Function.

How it works:
- You deploy a small Edge Function to your project
- The SDK calls the function with the source URL, bucket, and path
- The Edge Function fetches the file and streams it directly to Storage (server-side, single hop)

Benefits:
- No double bandwidth (file never touches your client)
- Includes SSRF protection out of the box
- Supports batch uploads with concurrency control

I'll share a link here once it's published. Would love feedback when it's posted and hopefully from the Supabase team on whether this approach could eventually be integrated into core.

Proposed examples:
1. Single upload
```js
import { createStorageHelpers } from '@supabase-community/storage-helpers'

const helpers = createStorageHelpers(supabase)

const result = await helpers.uploadFromUrl({
  bucket: 'images',
  path: 'ai-generated/image.png',
  sourceUrl: 'https://api.openai.com/v1/images/abc123.png',
})
```

2. Batch upload
```js
const results = await helpers.batchUploadFromUrl({
  bucket: 'images',
  jobs: [
    { sourceUrl: 'https://example.com/1.png', path: 'batch/1.png' },
    { sourceUrl: 'https://example.com/2.png', path: 'batch/2.png' },
    { sourceUrl: 'https://example.com/3.png', path: 'batch/3.png' },
  ],
  concurrency: 3,
})
```

**Comment by alexwhitmore** [comment]
The package should be ready to test, works pretty well! Let me know what you guys think

https://github.com/alexwhitmore/storage-helpers/tree/main
